// @cloudbase/node-sdk/dist/utils/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeConfig = exports.isValidEnvFormat = exports.isPageModuleName = exports.processReturn = exports.setThrowOnCode = exports.second = exports.isNonEmptyString = exports.E = exports.filterUndefined = exports.filterValue = exports.isAppId = exports.TcbError = void 0;
const utils = __importStar(require("../utils/utils"));
const code_1 = require("../const/code");
class TcbError extends Error {
    constructor(error) {
        super(error.message);
        this.code = error.code;
        this.message = error.message;
        this.requestId = error.requestId || '';
    }
}
exports.TcbError = TcbError;
function isAppId(appIdStr) {
    return /^[1-9][0-9]{4,64}$/gim.test(appIdStr);
}
exports.isAppId = isAppId;
function filterValue(o, value) {
    for (const key in o) {
        if (o[key] === value) {
            /* eslint-disable-next-line @typescript-eslint/no-dynamic-delete */
            delete o[key];
        }
    }
}
exports.filterValue = filterValue;
function filterUndefined(o) {
    filterValue(o, undefined);
}
exports.filterUndefined = filterUndefined;
function E(errObj) {
    return new TcbError(errObj);
}
exports.E = E;
function isNonEmptyString(str) {
    return typeof str === 'string' && str !== '';
}
exports.isNonEmptyString = isNonEmptyString;
function second() {
    // istanbul ignore next
    return Math.floor(new Date().getTime() / 1000);
}
exports.second = second;
// 兼容模式开关，兼容模式下，不抛出异常，直接返回
let throwOnCode = true;
function setThrowOnCode(value) {
    throwOnCode = value;
}
exports.setThrowOnCode = setThrowOnCode;
function processReturn(result) {
    if (!throwOnCode) {
        // 不抛报错，直接返回
        return result;
    }
    throw E(Object.assign({}, result));
}
exports.processReturn = processReturn;
/**
 * 是否是场景模块名
 *
 * $: 前缀，表示SaaS场景模块名，非实际环境ID，当前通过特殊环境ID标识
 *
 * @param envId
 * @returns
 */
function isPageModuleName(envId = '') {
    return typeof envId === 'string' && envId.startsWith('$:');
}
exports.isPageModuleName = isPageModuleName;
// 20 + 1 + 16, 限制长度 40
const kEnvRuleReg = /^[a-z0-9_-]{1,40}$/;
function isValidEnvFormat(env = '') {
    return typeof env === 'string' && kEnvRuleReg.test(env);
}
exports.isValidEnvFormat = isValidEnvFormat;
function normalizeConfig(config) {
    var _a, _b, _c, _d;
    const { debug = false, secretId, secretKey, env, timeout = 15000, headers = {}, accessKey } = config;
    // 检查用户显示配置的 secretId 和 secretKey 是否成对出现，要么都有，要么都没有
    if (!secretId !== !secretKey) {
        throw utils.E(Object.assign(Object.assign({}, code_1.ERROR.INVALID_PARAM), { message: 'secretId and secretKey must be a pair' }));
    }
    const cloudbaseConfigBase = Object.assign(Object.assign({}, config), { debug, 
        // secretId,
        // secretKey,
        envName: env, headers: Object.assign({}, headers), // 结构用户传进来的 headers，防止用户修改原对象
        timeout
        // accessKey: accessKey || process.env.CLOUDBASE_APIKEY
     });
    const { TENCENTCLOUD_SECRETID, TENCENTCLOUD_SECRETKEY, TENCENTCLOUD_SESSIONTOKEN, CLOUDBASE_APIKEY } = process.env;
    // 如果用户显示配置了accessKey，取用户显示配置的 accessKey,优先级最高
    if (accessKey) {
        return Object.assign(cloudbaseConfigBase, { secretId: undefined, secretKey: undefined, accessKey });
    }
    // 显示配置了 secretId, secretKey，取 secretId, secretKey
    if (secretId && secretKey) {
        return Object.assign(cloudbaseConfigBase, { secretId, secretKey, accessKey: undefined });
    }
    // 下面从环境变量取 secretId, secretKey, accessKey
    if (CLOUDBASE_APIKEY) {
        return Object.assign(cloudbaseConfigBase, { secretId: undefined, secretKey: undefined, accessKey: CLOUDBASE_APIKEY });
    }
    if (TENCENTCLOUD_SECRETID && TENCENTCLOUD_SECRETKEY) {
        return Object.assign(cloudbaseConfigBase, { secretId: TENCENTCLOUD_SECRETID, secretKey: TENCENTCLOUD_SECRETKEY, sessionToken: TENCENTCLOUD_SESSIONTOKEN, accessKey: undefined });
    }
    if ((_a = config.context) === null || _a === void 0 ? void 0 : _a.extendedContext) {
        const extendedContext = config.context.extendedContext;
        if (!cloudbaseConfigBase.env) {
            cloudbaseConfigBase.env = extendedContext.envId;
            cloudbaseConfigBase.envName = extendedContext.envId;
        }
        // 从 context 中获取 secret
        if (!cloudbaseConfigBase.secretId && !cloudbaseConfigBase.secretKey) {
            cloudbaseConfigBase.secretId = (_b = extendedContext === null || extendedContext === void 0 ? void 0 : extendedContext.tmpSecret) === null || _b === void 0 ? void 0 : _b.secretId;
            cloudbaseConfigBase.secretKey = (_c = extendedContext === null || extendedContext === void 0 ? void 0 : extendedContext.tmpSecret) === null || _c === void 0 ? void 0 : _c.secretKey;
            cloudbaseConfigBase.sessionToken = (_d = extendedContext === null || extendedContext === void 0 ? void 0 : extendedContext.tmpSecret) === null || _d === void 0 ? void 0 : _d.token;
        }
    }
    // 都没有配置，返回原始的配置
    return cloudbaseConfigBase;
}
exports.normalizeConfig = normalizeConfig;
